rclcpp/src/rclcpp/executor.cpp

```diff
@@ -108,20 +108,37 @@ Executor::spin_node_some(rclcpp::node::Node::SharedPtr node)
 void
 Executor::spin_some()
 {
-  while (AnyExecutable::SharedPtr any_exec =
-    get_next_executable(std::chrono::milliseconds::zero()))
-  {
+  if (spinning.exchange(true)) {
+    throw std::runtime_error("spin_some() called while already spinning");
+  }
+  AnyExecutable::SharedPtr any_exec;
+  while ((any_exec = get_next_executable(std::chrono::milliseconds::zero())) && spinning.load()) {
     execute_any_executable(any_exec);
   }
+  spinning.store(false);
```

Comment:
Got it.

I was trying to avoid adding mutexes to Executor where they're not needed. But `mutex` and `lock_guard` are the most suited to RAII-style management of this state. Maybe this would work with `try_lock` instead of `lock` so that we throw an exception if the Executor is already spinning instead of blocking.

@wjwwood, since you implemented about half of this PR, any thoughts?